### PR TITLE
non-node are accepted and converted as Text

### PR DESCRIPTION
**Pull request by caub**

### CODE REVIEW

File: polyfills/_mutation/polyfill.js [hunk 1]
```diff
@@ -1,22 +1,14 @@
...
-		var
-		fragment = document.createDocumentFragment(),
-		length = nodes.length,
-		index = -1,
-		node;
-
-		while (++index < length) {
-			node = nodes[index];
+function _mutation(nodes) {
+	if (nodes.length === 1) {
```
Comment (JakeChampion) [inline polyfills/_mutation/polyfill.js]: Do we need to cater for the case where nodes may be an empty array?

**Comment by JakeChampion** [comment]
Thanks for this PR as well @caub! Is it possible to have tests for this?

File: polyfills/_mutation/polyfill.js (hunk 1 above)
Comment (caub) [reply polyfills/_mutation/polyfill.js]: apparently no, it does nothing on chrome and the webplatform tests https://github.com/w3c/web-platform-tests/blob/master/dom/nodes/ChildNode-after.html#L14 allow it

for adding tests, could you point me in which folder please?

Comment (JakeChampion) [reply polyfills/_mutation/polyfill.js]: Ah, it looks like we have no tests for this polyfill at all :/

Comment (caub) [reply polyfills/_mutation/polyfill.js]: oh sorry I forgot tests are within each folder, bah for this 'helper' it doesn't matter so much

for .after, .before, and .replaceWith it'll need more tests

**Comment by JakeChampion** [comment]
Hi @caub, it looks like this PR has made some failing tests in IE7 and IE8:

```
Failures:
 - ie/8:
    -> Element.prototype.prepend
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.prepend
    -> Element.prototype.replaceWith
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.replaceWith
    -> Element.prototype.append
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.append
    -> Element.prototype.before
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.before
    -> Element.prototype.after
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.after
 - ie/7:
    -> Element.prototype.prepend
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.prepend
    -> Element.prototype.replaceWith
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.replaceWith
    -> Element.prototype.append
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.append
    -> Element.prototype.before
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.before
    -> Element.prototype.after
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.after
```

Do you think you would have some time to look into making these tests pass?

**Comment by caub** [comment]
Hi Jake,
locally the tests pass for me, until browserstack, but I have no keys for it.

I don't see why it would fail with the change, unless `Node` doesn't exists

ps: Did you consider passing tests like [web-platform tests](https://github.com/w3c/web-platform-tests) like in [jsdom](https://github.com/tmpvar/jsdom/blob/master/package.json#L78)? (by removing features before)

**Comment by JakeChampion** [comment]
@caub That is something we have considered for some time, yes. Do you have an idea of how we can get the web-platform-tests to work with this repo?

**Comment by JakeChampion** [comment]
@caub I made the test runner return more description failure messages:
```
Failures:
 - ie/8:
    -> Element.prototype.after can insert elements after itself
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.after
       'Node' is undefined
 - ie/8:
    -> Element.prototype.append can append elements to itself
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.append
       'Node' is undefined
 - ie/8:
    -> Element.prototype.before can insert elements before itself
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.before
       'Node' is undefined
 - ie/8:
    -> Element.prototype.prepend can prepend elements to itself
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.prepend
       'Node' is undefined
 - ie/8:
    -> Element.prototype.replaceWith can replace itself
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.replaceWith
       'Node' is undefined
 - ie/7:
    -> Element.prototype.after can insert elements after itself
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.after
       'Node' is undefined
 - ie/7:
    -> Element.prototype.append can append elements to itself
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.append
       'Node' is undefined
 - ie/7:
    -> Element.prototype.before can insert elements before itself
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.before
       'Node' is undefined
 - ie/7:
    -> Element.prototype.prepend can prepend elements to itself
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.prepend
       'Node' is undefined
 - ie/7:
    -> Element.prototype.replaceWith can replace itself
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.replaceWith
       'Node' is undefined
```

**Comment by caub** [comment]
Thanks, I did a way, not really beautiful (fallback to Element on ie8)

maybe polyfilling window.Node would be possible, but not easy

also _mutation is a global? couldn't we use module.exports there?

for web platform tests, I'll see, it's not easy because polyfill.io is not a DOM implementation, it just fills the holes, and browsers can be really different, so I guess browserstack is the best way

**Comment by JakeChampion** [comment]
@caub `_mutation` is a global to the polyfills but not global to user code as it is wrapped in a closure.

**Comment by JakeChampion** [comment]
@caub We're getting closer:
```
- ie/8:
    -> Element.prototype.after can insert elements after itself
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.after
       Function expected
 - ie/8:
    -> Element.prototype.append can append elements to itself
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.append
       Function expected
 - ie/8:
    -> Element.prototype.before can insert elements before itself
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.before
       Function expected
 - ie/8:
    -> Element.prototype.prepend can prepend elements to itself
       http://127.0.0.1:3000/test/tests?mode=targeted&feature=Element.prototype.prepend
       Function expected
```

File: polyfills/_mutation/polyfill.js [hunk 2]
```diff
@@ -1,22 +1,17 @@
...
-		}
+	if (nodes.length === 1) {
+		return nodes[0] instanceof _Node ? nodes[0] : document.createTextNode(nodes[0]+'');
+	}
+
+	var fragment = document.createDocumentFragment();
 
-		return fragment;
+	for (var i=0; i<nodes.length; i++) {
+		fragment.appendChild(nodes[i] instanceof _Node ? nodes[i] : document.createTextNode(nodes[i]+''));
```
Comment (JakeChampion) [inline polyfills/_mutation/polyfill.js]: `"text" instanceof Element` throws an error `"Function expected"`.

**Comment by JakeChampion** [comment]
@caub I added an isDOM helper function, do you think it is acceptable?

**Review by JakeChampion — APPROVED** [review]

**Comment by caub** [comment]
yes, I thought about checking .nodeType, etcc.. so ie8- don't expose those constructors, ok

**Pull request merged**